design/tls-backend-verification.md

```diff
@@ -0,0 +1,166 @@
+
+a. The DNS record may be hijacked, but without the matching certificate this is of little use.
+
+b. The certificate may be stolen, but the attacker must also redirect the DNS entries
+
+c. A certificate may be signed by an alternative CA, which is why pinning certificate fingerprints is a thing.
+
+The probability of a & b happening at the same time is non zero, but represent at least some effort to compensate for the trust issues of each component, this is why c is a thing.
+ 
+Let's compare this to how verification would work inside a Kubernetes cluster:
+Envoy, acting as the client, is configured to talk to a set of IP addresses obtained from k8s Endpoint objects--there are no hostnames in play inside a cluster--representing pods for that service.
+Envoy is configured by Contour to perform a TLS handshake when connecting to a pod.
+This design will add the following constraints to Envoy's TLS handshake.
+
+a. That the certificate presented is signed by the CA who's public key we present to Envoy via the IngressRoute `spec.routes[].backends[]verification.caSecret` key.
+
+b. That the certificate presented contains a subjectAltName that is present in the IngressRoute `spec.routes[].backends[].verification.subjectName` key.
+
+The problem is that both of these checks exist _inside the security boundary of the application which we are attempting to validate_.
+Because the certificate is almost certainly going to be signed by a company CA, not one of the public CAs, checking the CA presented by the server matches the material supplied to the client, Envoy, is little more than a shared secret.
+It does not assert that the certificate _is_ signed by a trusted CA, only that the certificate _is_ signed by a key to which the client has a matching public component.
+This is further weakened by the observation that the service which we are connecting too, the ingressroute, and the caSecret are in the _same namespace_.
+It is likely that the secret that the service we are connecting too is also in the same namespace.
+
+SubjectAltName verification is similarly diluted.
+There is no compensating control that the client will use the public DNS infrastructure to resolve the IP address to connect too as in the earlier example.
+Instead Envoy will connect to a set of IP addresses controlled by the namespace--Ingressroute and endpoint documents live in the same namespace--and ask the endpoint that answers what it's name is.
+SubjectAltName becomes just a shared secret, albeit weaker that the CA verification, because its just string matching, there's no crypto in there.
+
+To be clear, verification does offer some improvements over simply doing a TLS handshake, but I cannot convince myself that it is as secure as the public TLS infrastructure.
+The same CA, certificate, and subject name parameters can be reused across multiple clusters with exactly the same resultant validation as someone who went to the effort to issue a certificate per service.
```

Comment:
It's not really any less secure than public CA (as long as your private CA is managed properly), it's just not as obviously necessary as you have greater trust in your own private network. The two major benefits in this case are:
- Ensures that traffic doesn't get routed to the wrong pod even if there is a bug in Envoy
- Is the only way to be confident that your connection isn't subject to a MITM attack